src/inputs.rs

```diff
@@ -220,19 +220,34 @@ impl PartitionedInput {
                         let mut output_handle = output_wrapper.activate();
                         let mut output_session = output_handle.session(&output_cap);
                         for (key, part) in parts.iter() {
-                            let next = Python::with_gil(|py| part.next(py))
-                                .reraise_with(|| format!("error getting input for key {key:?}"));
-                            if let Some(mut items) = unwrap_any!(next) {
-                                // Only set just_emitted to false if all
-                                // the parts didn't emit an item.
-                                just_emitted = just_emitted | !items.is_empty();
-                                output_session.give_vec(&mut items);
-                                emit_keys_buffer.insert(key.clone());
-                            } else {
-                                tracing::trace!(
-                                    "Input {step_id_op:?} partition {key:?} reached EOF"
-                                );
-                                eofd_keys_buffer.insert(key.clone());
+                            // Ask the next awake time to the source.
+                            let next_awake =
+                                unwrap_any!(Python::with_gil(|py| part.next_awake(py)));
+                            let now = Utc::now();
+                            // Calculate for how long to wait for the next poll.
+                            activate_after = Some(
+                                next_awake
+                                    .signed_duration_since(now)
+                                    .max(min_cooldown)
```

Comment:
I would still keep the minimum cooldown because based on the benchmarks we conducted when implementing it, it not only reduced CPU usage but also (surprisingly) improved latency. Although I didn't delve deep into profiling to determine the exact reasons behind this, my impression was that the spin loop was stealing CPU time from other threads in the runtime, likely due to OS scheduling. If this is indeed the case, we could potentially address it by adjusting the niceness (priority) of threads. However, I'm hesitant to remove the cooldown without doing more specific benchmarks. If we want to optimize for minimum latency, I think it'd be better to work on that in a separate PR. Currently, removing the cooldown appears to have a negative impact based on the available data, so I would like to first create a precise benchmark and then make changes accordingly.

Regarding the larger issue you mentioned, are you suggesting that `next_awake` should return a `timedelta`, which would then be passed to the `activate_after` call, instead of performing the calculations on the Rust side? I did start implementing that approach initially, but I ended up using the `datetime` return value because it made the Python API easier to understand, without sacrificing any functionality. With the current implementation, if you want to activate the input source as soon as possible, you can simply return `datetime.now` (which is the default implementation), equivalent to `activate_after(min_cooldown)` as we were doing before. If you want to introduce a cooldown when there is no recent data, you can handle that in the `next` function and ensure that the next call to `next_awake` returns a datetime in the future. If you want to implement a metronome-like behavior, you can follow the example of PeriodicInput. Is there another case I'm not taking into account?

I believe it ultimately comes down to personal preference regarding the API rather than capabilities. You are correct that there is some implicit interaction between the layers. For example, we cannot guarantee with absolute certainty that `datetime.now` called from Python will always be strictly earlier than `Utc::now` called from Rust [a few moments later](https://doc.rust-lang.org/src/std/time.rs.html#159). However, since a negative wait value for us indicates the use of the minimum cooldown, it seems to me that providing lower-level control on the Python side does not offer users greater control over execution. On the contrary, users should make sure to not pass negative values, or it would fail when trying to convert to std::Duration. I opted for the current approach because it appeared to result in a much more intuitive API without sacrificing any capabilities.

By "more intuitive API" I mean that the system behaves as users would expect. They set a `next wakeup` time, and the next function is only called immediately after that moment has passed. If the user instructs us to activate after 10 seconds and we pass that value to `activate_after` without any other check, it will lead to unexpected activations, unless they are familiar with how Timely works and how Bytewax utilizes it. I'm not sure we should expect users to have knowledge of all those intricacies.